agent/connect/ca/provider_vault_auth.go

```diff
@@ -0,0 +1,239 @@
+	"github.com/hashicorp/go-cleanhttp"
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/go-secure-stdlib/awsutil"
+	"github.com/hashicorp/vault/api"
+	"github.com/hashicorp/vault/api/auth/gcp"
+
+	"github.com/hashicorp/consul/agent/structs"
+)
+
+// VaultAuthenticator defines the interface for logging into Vault using an auth method.
+type VaultAuthenticator interface {
+	// Login to Vault and return a Vault token.
+	Login(ctx context.Context, client *api.Client) (*api.Secret, error)
+}
+
+// LoginDataGenerator is used to generate the login data for a Vault login API request.
+type LoginDataGenerator interface {
+	// GenerateLoginData creates the login data from the given authMethod configuration.
+	GenerateLoginData(authMethod *structs.VaultAuthMethod) (map[string]interface{}, error)
+}
+
+var _ VaultAuthenticator = (*VaultAuthClient)(nil)
+
+// VaultAuthClient is a VaultAuthenticator that logs into Vault through the /auth/<method>/login API.
+type VaultAuthClient struct {
+	// AuthMethod holds the configuration for the Vault auth method login.
+	AuthMethod *structs.VaultAuthMethod
+	// LoginPath is optional and can be used to explicitly set the API path that the client
+	// will use for a login request. If it is empty the path will be derived from AuthMethod.MountPath.
+	LoginPath string
+	// LoginDataGen derives the parameter map containing the data for the login API request.
+	// For some auth methods this is needed to transform the AuthMethod.Params into the login data
+	// needed for the API request.
+	LoginDataGen LoginDataGenerator
+}
+
+// NewVaultAPIAuthClient creates a VaultAuthClient that uses the Vault API to perform a login.
+func NewVaultAPIAuthClient(authMethod *structs.VaultAuthMethod, loginPath string) *VaultAuthClient {
+	if loginPath == "" {
+		mount := authMethod.Type
+		if authMethod.MountPath != "" {
+			mount = authMethod.MountPath
+		}
+		loginPath = fmt.Sprintf("auth/%s/login", mount)
+	}
+	return &VaultAuthClient{
+		AuthMethod: authMethod,
+		LoginPath:  loginPath,
+	}
+}
+
+// Login performs a Vault login operation and returns the associated Vault token.
+func (c *VaultAuthClient) Login(ctx context.Context, client *api.Client) (*api.Secret, error) {
+	var err error
+	loginData := c.AuthMethod.Params
+
+	// If a login data generator is provided then use it to transform the auth method params
+	// into the appropriate login data for the API request.
+	if c.LoginDataGen != nil {
+		loginData, err = c.LoginDataGen.GenerateLoginData(c.AuthMethod)
+		if err != nil {
+			return nil, fmt.Errorf("aws auth failed to generate login data: %w", err)
+		}
+	}
+
+	if ctx == nil {
+		ctx = context.Background()
+	}
+	return client.Logical().WriteWithContext(ctx, c.LoginPath, loginData)
+}
+
+var _ VaultAuthenticator = (*gcp.GCPAuth)(nil)
+
+// NewGCPAuthClient returns a VaultAuthenticator that can log into Vault using the GCP auth method.
+func NewGCPAuthClient(authMethod *structs.VaultAuthMethod) (VaultAuthenticator, error) {
```

Comment:
nit: It would be nice to put the GCP stuff here in its own file, and maybe the AWS stuff in its own file. I guess, `provider_vault_auth_gcp.go`, `provider_vault_auth_aws.go`? What do you think?